Enable multiple parameters in `im_prg`, override shebang line with im_prg if set

pbsds opened the pull request:
This PR does the following:

* Split `im_opt` with `shlex.split` instead of str.split
* Split im_prg with `shlex.split`, allowing passing multiple options in im_prg
* Make it so `im_prg` optionally overrides the shebang line in `.shebang`

## Request for comment:

`Handler.im_prg` has with this PR changed from a `str` to a `List[str]`. Will this break anything?
I've looked over the worker dispatch logic, and i believe the `klass` is set by the matched pandoc codeblock class and not im_prg. I'm not sure how to properly test this.

hertogp commented:
Hi @pbsds,

On the surface, it doesn't appear to break anything, but I tested only with mermaid, boxes & shebang.  Would need more time to test the rest.  If you want to test, you can create a new  markdown file, e.g. `nice.md` in the example directory and run `make nice`.  For example, I cloned your fork, checked out the `im_prg` branch and in the example sub-dir, I created a `nice.md` file: 

---
    $ cat nice.md
    ---
    imagine.shebang.im_out: stdout,fcb
    imagine.im_log: 4
    ...
    
    # Nice unicorn
    
    ```{.shebang im_prg="nice -n 10"}
    #!/bin/bash
    cat << 'EOF' | boxes -d peek -ph4v1
    Never agree with stupid people,
    they'll drag you down to their level
    and then beat you with experience.
                          -- Mark Twain
    EOF
    ```
---

and ran

---
    $ make nice
    ---------  nice  ---------
    /usr/bin/pandoc --filter ../pandoc_imagine.py nice.md defaults.yaml -o nice.pdf
    Imagine[SheBang  :debug] exec:  nice -n 10 pd-images/0ba4765074af176c2fa6177b19d1a7558cfd44be.shebang pd- images/0ba4765074af176c2fa6177b19d1a7558cfd44be.png
    Imagine[SheBang  : note] stderr> saw 0 bytes
    Imagine[SheBang  :debug] stdout> /*       _\|/_
    Imagine[SheBang  :debug] stdout>          (o o)
    Imagine[SheBang  :debug] stdout>  +----oOO-{_}-OOo-----------------------------+
    Imagine[SheBang  :debug] stdout>  |                                            |
    Imagine[SheBang  :debug] stdout>  |    Never agree with stupid people,         |
    Imagine[SheBang  :debug] stdout>  |    they'll drag you down to their level    |
    Imagine[SheBang  :debug] stdout>  |    and then beat you with experience.      |
    Imagine[SheBang  :debug] stdout>  |                          -- Mark Twain     |
    Imagine[SheBang  :debug] stdout>  |                                            |
    Imagine[SheBang  :debug] stdout>  +-------------------------------------------*/
    Imagine[SheBang  : note] stdout> saw 414 bytes
---

where

---
    $ cat default.yaml  # gives:
    ---
    imagine.im_out: img,fcb
    imagine.im_log: 4
    ...
---

In the current setup, a subclass `X(Handler)` registers itself for all the keys in its `cmdmap`.  So a codeblock with `some-class`
is handled by the `X(Handler)` that registered `some-class` by having an (key-)entry for `some-class` in its `cmdmap` pointing to the cli-program that should handle the codeblock's contents . That cli-program can be overridden by supplying the `im_prg` option, either in the codeblock itself, or in the doc's meta-data (or in a defaults.yaml-file \<-- need to double-check that!).

Currently the command-line to be executed by `Popen` is built by (`im_prg`, `*args`), where there is no possibility to re-route the execution of a system-script (like a `shebang`) through a dispatcher like `nice` with options of its own.

However, I think a new handler class would be cleaner since that would not touch the rest of the code base and explicit is better
than implicit, or so the saying goes.

As an example, in my own `via`-branch, I added an new SheBangVia class:

---
    class SheBangVia(Handler):
    '''
    http://www.google.com/search?q=shebang+line
    '''
    # runs fenced code block via `im_prg` to be handled as a system script'
    cmdmap = {'shebangvia': ''}
    
    def image(self):
        '[im_prg] <fname>.shebangvia {im_opt} <fname>.{im_fmt}'
        os.chmod(self.inpfile, stat.S_IEXEC | os.stat(self.inpfile).st_mode)
        args = [self.inpfile] + self.im_opt + [self.outfile]
        via = self.im_prg.split()
        if self.cmd(*via, *args):
            return self.result()
---

and, in the example directory, a `via.md` wth:

---
    ---
    imagine.shebangvia.im_log: 4
    imagine.shebangvia.im_out: stdout,fcb
    ...

    # Standard unicorn

    ```shebangvia
    #!/bin/bash
    cat << 'EOF' | boxes -d peek -ph4v1
    Never agree with stupid people,
    they'll drag you down to their level
    and then beat you with experience.
                          -- Mark Twain
    EOF
    ```

    # Via Nice Unicorn

    ```{.shebangvia im_prg="nice -n 10"}
    #!/bin/bash
    cat << 'EOF' | boxes -d peek -ph4v1
    Never agree with stupid people,
    they'll drag you down to their level
    and then beat you with experience.
                          -- Mark Twain
    EOF
    ```
---

and ran

---
    $ make via

    ---------  via  ---------
    /usr/bin/pandoc --filter ../pandoc_imagine.py via.md defaults.yaml -o via.pdf
    Imagine[SheBangVia:debug] exec:  pd-images/bc75ae639f5a71d377d3f4a610e2941119ecef87.shebangvia pd- images/bc75ae639f5a71d377d3f4a610e2941119ecef87.png
    Imagine[SheBangVia: note] stderr> saw 0 bytes
    Imagine[SheBangVia:debug] stdout> /*       _\|/_
    Imagine[SheBangVia:debug] stdout>          (o o)
    Imagine[SheBangVia:debug] stdout>  +----oOO-{_}-OOo-----------------------------+
    Imagine[SheBangVia:debug] stdout>  |                                            |
    Imagine[SheBangVia:debug] stdout>  |    Never agree with stupid people,         |
    Imagine[SheBangVia:debug] stdout>  |    they'll drag you down to their level    |
    Imagine[SheBangVia:debug] stdout>  |    and then beat you with experience.      |
    Imagine[SheBangVia:debug] stdout>  |                          -- Mark Twain     |
    Imagine[SheBangVia:debug] stdout>  |                                            |
    Imagine[SheBangVia:debug] stdout>  +-------------------------------------------*/
    Imagine[SheBangVia: note] stdout> saw 414 bytes
    Imagine[SheBangVia:debug] exec:  nice -n 10 pd-images/1ea6999414aee636e72ee0496f33fe49b5d8bed8.shebangvia pd- images/1ea6999414aee636e72ee0496f33fe49b5d8bed8.png
    Imagine[SheBangVia: note] stderr> saw 0 bytes
    Imagine[SheBangVia:debug] stdout> /*       _\|/_
    Imagine[SheBangVia:debug] stdout>          (o o)
    Imagine[SheBangVia:debug] stdout>  +----oOO-{_}-OOo-----------------------------+
    Imagine[SheBangVia:debug] stdout>  |                                            |
    Imagine[SheBangVia:debug] stdout>  |    Never agree with stupid people,         |
    Imagine[SheBangVia:debug] stdout>  |    they'll drag you down to their level    |
    Imagine[SheBangVia:debug] stdout>  |    and then beat you with experience.      |
    Imagine[SheBangVia:debug] stdout>  |                          -- Mark Twain     |
    Imagine[SheBangVia:debug] stdout>  |                                            |
    Imagine[SheBangVia:debug] stdout>  +-------------------------------------------*/
    Imagine[SheBangVia: note] stdout> saw 414 bytes
---

The first 'image' has no `im_prg` and is handled as a regular `shebang` while the second is run through `nice` with param `-n 10`.

Could you explore that approach and see if it fits your use-case(s)?

Thanks!

pbsds commented:
This is a much simpler approach IMO, and should fit my use-case.
There is the edge case however where a user specifies `.shebangvia` without setting `im_prg`.
I would either return a `fcb` telling the user what went wrong, or log a message for all non-negative logging levels.

I encourage using `shlex.split` instead of `str.split` in `shebangvia` and `im_opt`. 
Consider cherry-picking the first commit in this PR.

hertogp commented:
Could you checkout the `via` branch and the `examples/imprg.md` to see if it does the job?

Note: as class name, I landed on `ImPrg` which seems better than `SheBangVia`.

btw. thanks for pointing out `shlex.split` as a better alternative to a mere `"string".split()`

pbsds commented:
Sorry for my late reply, I'm crunching on my thesis due in June.

`ImPrg` is perhaps a more intuitive name, as the shebang line could be omitted by referencing an interpreter directly.

    ```{.imprg im_prg="python" im_out="stdout"}
    print("Hello world, I ran without a shebang line!")
    ```

However, now the behavior of setting the executable flag on the resulting `*.imprg` file is not immediately obvious to the user.
Python does not require the executable flag being set, but `nice` does. 
It is therefore important to include the ImPrg class in the security chapter in the readme.

I looked over your branch, and it should work as it splits the `im_prg`. I have not run it, as i'm sticking to my fork until I deliver.

pbsds commented:
Posting this as a separate comment, as it is a bit of-topic. A small insight into my insane use-cases and workflow:

I combine pandoc-imagine with pantable, generating my tables with a python script and then rendering them with pantable:

    ```{.python .table .imprg im_prg="poetry run python" im_out="stderr"}
    if 1: # Allows me to indent the code, and fold it in my editor
        # i use the .python class for proper highlighting in my editor
        # i use .imprg to run this via pandoc-imagine
        # i then use .table to convert the output to a pandoc table via pantable
        # i use stderr instead of stdout, because poetry litters the stdout with junk. I've logged an issue
        row = lambda *x: print(*x, sep=",", file=sys.stderr) # helper function
        row("#", "a", "b") # header
        for i in range(1, 3+1):
            row(i, i*2, i**2)
    ```

Here I notice how pandoc-imagine do "cache" generated images, but not generated stdout not stderr.
This causes the script above to be run every time compile the document.

I also render pretty hefty 3D models with matplotlib. Generating those models require access to a GPU with CUDA, taking upwards of 8 minutes for a single image.
I therefore run pandoc via [remote](https://github.com/remote-cli/remote) on a remote host, downloading the resulting `pd-images` folder with rsync. On the remote host i ran the python script with `#!/usr/bin/env -S poetry run python`, but ran into the issue with coreutils being too old to support `-S`, as the remote host runs ubuntu 18.04.
